### ISSUE TITLE
Consider warning by default for `foo == bar;` statements

### ISSUE THREAD

**Issue by crumblingstatue** [rust-lang/rust#38224]
One can make a mistake where they mistype `=` to `==`.

Consider the following code:
``` Rust
fn main() {
    let mut a = 10;
    println!("A is {}", a);
    if 1 == 2 {
        a = 20;
    } else {
        a == 30; // Oops, meant assignment
    }
    println!("A is now {}", a); // Still 10
}
```
No warning is produced for this code.

I'd argue that this is a common enough mistake to justify warning by default for it.

If the user explicitly wants to do this operation for side effects or whatever, they can do `let _ = a == b;`, similarly to the explicit version for `#[must_use]` types.

**Comment by mbrubeck** [comment]
I think this is a good candidate for an on-by-default builtin lint:

* It catches a typo that is easy to make and difficult to spot, and that won't be caught by type checking.
* It can be very narrowly targeted, only matching statements of the form `EXPR == EXPR;`.
* False positives are unlikely, because `==` should rarely if ever have side effects, so it almost never makes sense to discard its result.

**Comment by crumblingstatue** [comment]
> False positives are unlikely, because == should rarely if ever have side effects, so it almost never makes sense to discard its result.

We could also show a hint to use the more explicit `let _ = a == b` version in case the user wants to only evaluate side effects.

**Comment by nagisa** [comment]
> I think this is a good candidate for an on-by-default builtin lint

Needs RFC.

**Comment by crumblingstatue** [comment]
Closing in favor of https://github.com/rust-lang/rfcs/pull/1812.
